frontend/src/metabase/metadata/components/NameDescriptionInput/NameDescriptionInput.tsx

```diff
@@ -0,0 +1,113 @@
+import { useLayoutEffect, useState } from "react";
+
+import { Flex, Icon, TextInput } from "metabase/ui";
+
+import S from "./NameDescriptionInput.module.css";
+
+interface Props {
+  description: string;
+  descriptionPlaceholder: string;
+  name: string;
+  namePlaceholder: string;
+  onDescriptionChange: (description: string) => void;
+  onNameChange: (name: string) => void;
+}
+
+/**
+ * Controlled component that fires on*Change events on blur
+ */
+export const NameDescriptionInput = ({
+  description,
+  descriptionPlaceholder,
+  name,
+  namePlaceholder,
+  onDescriptionChange,
+  onNameChange,
+}: Props) => {
+  const [nameState, setNameState] = useState(name);
+  const [descriptionState, setDescriptionState] = useState(description);
+  const [isNameFocused, setIsNameFocused] = useState(false);
+  const [isNameHovered, setIsNameHovered] = useState(false);
+  const [isDescriptionFocused, setIsDescriptionFocused] = useState(false);
+  const [isDescriptionHovered, setIsDescriptionHovered] = useState(false);
+
+  useLayoutEffect(() => {
+    setNameState(name);
+  }, [name]);
+
+  useLayoutEffect(() => {
+    setDescriptionState(description);
+  }, [description]);
+
+  return (
+    <div>
+      <TextInput
+        classNames={{
+          input: S.nameInput,
+          root: S.name,
+        }}
+        fw="bold"
+        placeholder={namePlaceholder}
+        rightSection={
+          isNameHovered && !isNameFocused ? (
+            <Flex className={S.rightSection}>
+              <Icon name="pencil" size={12} />
+            </Flex>
+          ) : undefined
+        }
+        rightSectionPointerEvents="none"
+        size="lg"
+        value={nameState}
+        onBlur={(event) => {
+          setIsNameFocused(false);
+
+          const newValue = event.target.value;
+
+          // prevent empty names
+          if (!newValue.trim()) {
+            setNameState(name);
```

Comment:
Could this be pulled out of the if block? Either way is probably fine, but it simplifies the code flow a bit.

Reply:
> Could this be pulled out of the if block?

Not sure how due to `return;`.

I'll pull it out of jsx block though.